File: frontend/src/ui/domain/week-grid/readings.ts
/* THE STRIP'S THREE FIGURES, AND THE WORD THAT QUALIFIES THE BLOCK COUNT.
 *
 * A figure arrives as minutes and reads as hours, and how it reads is a domain decision rather than a container's,
 * which is why `StatCell` takes a formatted figure and this module produces one.
 *
 * ONE DECIMAL PLACE. `80.8h` is what the design record renders and it is the right precision for a week: a whole
 * number hides a two-hour swing across three cells, and two decimals invite a reader to compare minutes on a
 * surface whose whole point is the shape of a week.
 *
 * PLAN CURRENCY SHARES THE SCHEDULED CELL'S SUB-LINE rather than claiming a fifth cell, because the strip's
 * content budget is settled at three readings plus the verdict, and currency qualifies the block count, which is
 * exactly what it does: it says whether that count is current. Never a spinner, and never anything that moves.
 *
 * THE SHARE IS A RATIO OF THE TWO FIGURES THE SERVER SENT, never a difference computed here. The same arithmetic
 * produces both, so the strip and the pie review cannot disagree about a percentage even while the denominator's
 * own definition is under revision. Deriving it from what the grid happens to draw would make two surfaces answer
 * one question differently, which is the whole reason the figures are computed server-side. */

/** Whether the block count is current, being recomputed, or the last one that worked. Never a spinner. */
export type PlanCurrency = "current" | "solving" | "stale";

/** The three readings, the count they describe, and the word that qualifies it. */
export interface StripReadings {
  readonly scheduledMinutes: number;
  readonly discretionaryMinutes: number;
  readonly unallocatedMinutes: number;
  readonly blockCount: number;
  readonly planCurrency: PlanCurrency;
}

const MINUTES_IN_HOUR = 60;
const PLACES = 1;
const PERCENT = 100;

/** Minutes as hours, to one decimal place: `80.8h`. */
export function formatHours(minutes: number): string {
  return `${(minutes / MINUTES_IN_HOUR).toFixed(PLACES)}h`;
}

/**
 * A share of the discretionary denominator: `35.3% of discretionary`.
 *
 * A zero denominator reads as the sentence rather than as `NaN%` or `0.0%`: a week with no discretionary time at
 * all has no share to report, and reporting zero would say the opposite of what is true.
 */
export function formatShare(minutes: number, ofMinutes: number): string {
  if (ofMinutes <= 0) return "no discretionary time";
  return `${((minutes / ofMinutes) * PERCENT).toFixed(PLACES)}% of discretionary`;
}

/**
 * The SCHEDULED cell's sub-line: the block count, and the word that says whether it is current.
 *
 * `91 blocks` when it is, `91 · solving` and `91 · stale` when it is not. The count is spelled out only in the
 * current case, because the other two need the room for the word and the word is what the reader is being told.
 */
export function formatCurrency(blockCount: number, currency: PlanCurrency): string {
  if (currency === "current") return `${blockCount} blocks`;
  return `${blockCount} · ${currency}`;
}

File: packages/syncr-api/src/syncr_api/plans/readings.py
"""The nine figures the summary strip and the review both read, computed once on the server.

The strip shows three of them plus the verdict, and the pie review divides the same denominator.
Computing them in a client from the document would risk a figure on the strip disagreeing with the
same figure in a review, so they are computed here and only here.

## Where each figure comes from, and why the four budget ones are not the document's

``discretionary``, ``unallocated``, ``oversubscription`` and ``off_plan`` are taken from the BUDGET
REPORT, which is the one implementation of that arithmetic and the one the pie review divides. The
document carries three of them too, computed when the plan was produced, and those are as of that
instant: a period declared off-plan an hour later moves the real denominator and not the stored one.
Reading the report is therefore what makes the strip and the review agree by construction rather
than by a test, and it is why the wire document does not carry the stored three at all.

``scheduled``, ``block_count`` and ``unconfirmed_days`` are the document's, because no other shape
can answer them: they are facts about what the week holds. ``dropped_legs`` is the document's
for the same reason, and a figure rather than something a client counts out of the payload:
an absence explained on the grid is still one fact about the week.

## Three of the nine are worth stating exactly

``scheduled_minutes`` is a COVERAGE figure, unioned and clipped to the week. Unioned because a
minute the user deliberately double-booked is one scheduled minute rather than two, and clipped
because a Sunday-night routine running into Monday is scheduled time that belongs to next week.

``unconfirmed_days`` counts only the week's days that have ENDED, and only those holding a block. A
day the user cannot confirm yet is not outstanding, and a day with nothing on it has nothing to
confirm: counting either would make the figure nag about work nobody owes.

``plan_currency`` is derived from the week's operation state, in ``currency.py``, so the strip's
sub-line and the operation resource cannot disagree.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

from syncr_domain.gaps import EmptySlotReason
from syncr_domain.intervals import IntervalSet
from syncr_domain.weeks import local_days

if TYPE_CHECKING:
    from collections.abc import Collection
    from datetime import datetime

    from syncr_api.offplan.reading import OffPlanReading
    from syncr_api.plans.currency import PlanCurrency
    from syncr_domain.budgets import BudgetReport
    from syncr_domain.intervals import Interval
    from syncr_domain.plan import PlanDocument
    from syncr_domain.zones import Date


@dataclass(frozen=True, slots=True)
class WeekReadings:
    """The strip's three readings, its sub-line, and the four figures beside them."""

    scheduled_minutes: int
    discretionary_minutes: int
    unallocated_minutes: int
    oversubscription_minutes: int
    unconfirmed_days: int
    off_plan_minutes: int
    block_count: int
    dropped_legs: int
    plan_currency: PlanCurrency


def week_readings(
    document: PlanDocument,
    *,
    span: Interval,
    report: BudgetReport,
    off_plan: OffPlanReading,
    confirmed: Collection[Date],
    now: datetime,
    currency: PlanCurrency,
) -> WeekReadings:
    """Every reading for one week that holds a plan."""
    return WeekReadings(
        scheduled_minutes=scheduled_minutes(document, span),
        discretionary_minutes=report.discretionary_minutes,
        unallocated_minutes=report.unallocated_minutes,
        oversubscription_minutes=report.oversubscription_minutes,
        unconfirmed_days=unconfirmed_days(document, span=span, confirmed=confirmed, now=now),
        off_plan_minutes=off_plan.minutes,
        block_count=len(document.blocks),
        dropped_legs=sum(
            1 for slot in document.empty_slots if slot.reason is EmptySlotReason.DROPPED_LEG
        ),
        plan_currency=currency,
    )


def scheduled_minutes(document: PlanDocument, span: Interval) -> int:
    """Minutes of the week that hold a block, counting a double-booked minute once."""
    return IntervalSet(block.interval for block in document.blocks).clip(span).total_minutes()


def unconfirmed_days(
    document: PlanDocument, *, span: Interval, confirmed: Collection[Date], now: datetime
) -> int:
    """The week's ended days that hold a block and that the user has not confirmed.

    The days are bounded by the document's own zone mapping, so a date is as long as it really was:
    a spring-forward date is 23 hours and a travel boundary makes one 14. That mapping was captured
    when the plan was produced, which is also what the grid draws the day columns from, so the day a
    figure is charged to is the day the reader sees.

    **The span is live and the mapping is stored, and they can disagree.** A travel override
    declared after the plan was produced moves the week's real bounds relative to the dates the
    document captured. ``local_days`` clips every day to the span and drops one whose bounds do not
    run forward, so the disagreement costs a day rather than a fault, and charging the figure to the
    day the grid draws is the reading a person can check.
    """
    return sum(
        1
        for day in local_days(document.iso_week, document.zone_by_date, span)
        if day.interval.end <= now
        and day.on not in confirmed
        and any(block.interval.overlaps(day.interval) for block in document.blocks)
    )
